### ISSUE TITLE
Make all async functions return a Future

### ISSUE THREAD

**Issue by synw** [centrifugal/centrifuge-dart#31]
Would it be possible to return a future for all async methods? I have noticed that some async methods return `void` and not `Future<void>` which make then non awaitable. Is this on purpose?

Ex: I would like to await `connect` but  can't because it does not return a future:

```dart
@override
  void connect() async {
    return _connect();
  }
```

Here `_connect` returns a future, and `connect` would be great this way:

 ```dart
  Future<void> connect() async {
    return await _connect();
  }
// or the short way
  Future<void> connect() async => _connect();
```

Is it possible or the actual way is by design?

**Comment by FZambia** [comment]
@synw hello, the reason is that there are streams with events which can happen many times during client lifetime. For example `client.connectStream`. If you call `connect` then I suppose you expect it to yield when connection successfully established or an error occurred. But only for the first time right?

But you can also want to do some work as soon as connection established (write sth to a screen). But suppose you also need to do this work after each reconnect. In this case you also register your work in connectStream handler. In this case you need a way to only do your work once. In general we can solve this by adding `isReconnect` flag to connect context but checking it seems a bit non-obvious thing for me.

I don't have strong opinion here, maybe you can describe an API you want with some examples and use cases? With understanding that events like `connect`, `disconnect`, `subscribeSuccess`, `unsubscribe` can happen many times throughout client lifetime. For me event model looks more natural to deal with but I can miss some scenarios.

**Comment by synw** [comment]
> I don't have strong opinion here, maybe you can describe an API you want with some examples and use cases?

Ok. The non awaitable `connect` was blocking for me when trying to implement a reconnect with my api experiments, trying to manage things manually somehow. I'll try to post more info about the use case if I dig more this

Please notice that an async function returning void is non idiomatic in Dart: there is a [linting rule](https://dart-lang.github.io/linter/lints/avoid_void_async.html) about this. 

Btw I would recommend using more strict analysis options for this package. [Pedantic](https://github.com/dart-lang/pedantic) is quite standard now as used by Google and for pub.dev packages evaluation ([pana](https://pub.dev/packages/pana)). I personally use and highly  recommend [Extra Pedantic](https://github.com/modulovalue/extra_pedantic), which is very strict. You already use `implicit-dynamic: false` in analysis options and it's good. Maybe you could add `implicit-casts: false` as well: it helps preventing bugs by being less cool about types

**Comment by FZambia** [comment]
@synw thanks for suggestions. Since I maintain all Centrifuge libs and Dart is not my day to day language I hardly can find time and knowledge to refactor for idiomatic code. If you are using this library then maybe you can send pr with all your code style recommendations - will be happy to accept.

**Comment by SimonHausdorf** [comment]
I was looking for an await connection / await subscription, too.

1. initiate the web socket connection, on success
2. call an API endpoint to fetch the current value of something.

Let's say due to the delay of connection, I call the API endpoint first. I receive `value = true`. Now the backend is publishing `value = false` but the connection gets established after this.

Now my value remains true but in the backend it already changed to false.

I know that there are some ways to solve the issue as example working with history and receiving the last messages from the channel, however, I believe that this is not optimal.

**Comment by FZambia** [comment]
@pashtetbezd hi, your question is off topic here. Try asking in community rooms. General answer - you cant use websockets in background - you should disconnect and use push notifications for important messages.

**Comment by pashtetbezd** [comment]
I wanted to say that if the function returned futures there would be no question caused streams hell. Please add the futures.

**Comment by FZambia** [comment]
Well, addressing this in #57 – since everybody asks about this, let's try and see how it goes. If anyone can test out branch from #57 – plz do and share feedback.

**Comment by FZambia** [comment]
Addressed by [v0.8.0](https://github.com/centrifugal/centrifuge-dart/releases/tag/v0.8.0)
